tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py

```diff
@@ -78,9 +79,92 @@ class PortForwardingError(Exception):
+                 namespace: str,
+                 destination: str,
+                 remote_port: int,
+                 local_port: Optional[int] = None,
+                 local_address: Optional[str] = None):
+        self.context = context
+        self.namespace = namespace
+        self.destination = destination
+        self.remote_port = remote_port
+        self.local_address = local_address or self.PORT_FORWARD_LOCAL_ADDRESS
+        self.local_port, self.subprocess = self._establish_port_forwarding(
+            local_port)
+
+    def _establish_port_forwarding(self, local_port: Optional[int] = None):
+        port_mapping = f"{local_port}:{self.remote_port}" if local_port else f":{self.remote_port}"
+        cmd = [
+            "kubectl", "--context", self.context, "--namespace", self.namespace,
+            "port-forward", "--address", self.local_address, self.destination,
+            port_mapping
+        ]
+        pf = subprocess.Popen(cmd,
+                              stdout=subprocess.PIPE,
+                              stderr=subprocess.STDOUT,
+                              universal_newlines=True)
+        # Wait for stdout line indicating successful start.
+        if local_port:
+            local_port_expected = (
+                f"Forwarding from {self.local_address}:{local_port}"
+                f" -> {self.remote_port}")
+        else:
+            local_port_re = re.compile(
+                f"Forwarding from {self.local_address}:([0-9]+) -> {self.remote_port}"
+            )
+        try:
+            while True:
+                time.sleep(0.05)
+                output = pf.stdout.readline().strip()
+                if not output:
+                    return_code = pf.poll()
+                    if return_code is not None:
+                        errors = [error for error in pf.stdout.readlines()]
+                        raise PortForwardingError(
+                            'Error forwarding port, kubectl return '
+                            f'code {return_code}, output {errors}')
+                    # If there is no output, and the subprocess is not exiting,
+                    # continue waiting for the log line.
+                    continue
+
+                # Validate output log
+                if local_port:
+                    if output != local_port_expected:
+                        raise PortForwardingError(
+                            f'Error forwarding port, unexpected output {output}'
+                        )
+                else:
+                    groups = local_port_re.search(output)
+                    if groups is None:
+                        raise PortForwardingError(
+                            f'Error forwarding port, unexpected output {output}'
+                        )
+                    local_port = int(groups[1])
+
+                logger.info(output)
+                break
+        except Exception:
+            self.port_forward_stop(pf)
```

Comment:
```suggestion
            self.close()
```

You'd need to set `self.subprocess = pf` too

Reply:
Done, with the above `connect` change, this function now is a state mutating method.